spec/support/matchers/match_collection.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+RSpec::Matchers.define :match_collection do |expected_collection, attribute_names: []|
+  match do |actual_collection|
+    @expected_collection = extract_debug_info(expected_collection, attribute_names)
+    @actual_collection = extract_debug_info(actual_collection, attribute_names)
+
+    @expected_collection == @actual_collection
+  end
+
+  failure_message do
+    missing_elements = diff_collections(@expected_collection, @actual_collection, :missing)
+    extra_elements = diff_collections(@expected_collection, @actual_collection, :extra)
+
+    message = <<~MESSAGE
+      #{'Expected collection contained:'.colorize(:yellow)}
+        #{@expected_collection.pretty_inspect.colorize(:green)}
+      #{'Actual collection contained:'.colorize(:yellow)}
+        #{@actual_collection.pretty_inspect.colorize(:red)}
+    MESSAGE
+
+    message += <<~MESSAGE if missing_elements.any?
+      #{'The missing elements were:'.colorize(:yellow)}
+        #{missing_elements.pretty_inspect.colorize(:blue)}
+    MESSAGE
+
+    message += <<~MESSAGE if extra_elements.any?
+      #{'The extra elements were:'.colorize(:yellow)}
+        #{extra_elements.pretty_inspect.colorize(:magenta)}
+    MESSAGE
+
+    message
+  end
+
+  def extract_debug_info(collection, attribute_names)
+    collection.map do |item|
+      attribute_names.each_with_object({}) do |attr, hash|
+        value = item.public_send(attr) if item.respond_to?(attr)
+        hash[attr] = value
+      end.merge(id: item.id)
+    end
+  end
+
+  def diff_collections(expected, actual, type)
+    expected_ids = expected.pluck(:id)
+    actual_ids = actual.pluck(:id)
+
+    case type
+    when :missing
+      expected.reject { |e| actual_ids.include?(e[:id]) }
+    when :extra
+      actual.reject { |a| expected_ids.include?(a[:id]) }
+    end
+  end
+end
```

Comment:
I created this matcher because it was impossible to know what is wrong on the spec:

```sh
1) CoursesQuery.call when filter for visa sponsorship returns courses that sponsor visa
     Failure/Error: expect(results).to contain_exactly(course_that_sponsor_visa, another_course_that_sponsor_visa)

                 Expected [#<Course:0x0000000135c1aa08 @_already_called=nil, @_committed_already_called=nil, @_defer_touch_attrs=nil, @_new_record_before_last_commit=nil, @_skip_dirty_tracking=nil, @_start_transaction_state=nil, @_touch_attr_names=nil, @_touch_record=nil, @_trigger_destroy_callback=nil, @_trigger_update_callback=nil, @association_cache={ course_subjects: #<ActiveRecord::Associations::HasManyAssociation:0x0000000143b34

1000000 lines long.
```

Reply:
Now is much better:

```
Expected collection contained:
         [{"can_sponsor_skilled_worker_visa"=>true,
         "can_sponsor_student_visa"=>false,
         :id=>4},
        {"can_sponsor_skilled_worker_visa"=>false,
         "can_sponsor_student_visa"=>true,
         :id=>5}]

       Actual collection contained:
         [{"can_sponsor_skilled_worker_visa"=>true,
         "can_sponsor_student_visa"=>false,
         :id=>4},
        {"can_sponsor_skilled_worker_visa"=>false,
         "can_sponsor_student_visa"=>true,
         :id=>5},
        {"can_sponsor_skilled_worker_visa"=>false,
         "can_sponsor_student_visa"=>false,
         :id=>6}]


       The missing elements were:
         []


       The extra elements were:
         [{"can_sponsor_skilled_worker_visa"=>false,
         "can_sponsor_student_visa"=>false,
         :id=>6}]
```